### PR TITLE
Remove vault dependency

### PR DESCRIPTION
**Pull request by shydefoo**
### Motivation
* This PR removes Turing's dependency on vault. 
* Access to clusters is replaced with a truncated form of a kubeconfig. [To explain in PR]
* Depends on this https://github.com/gojek/mlp/pull/74 in gojek/mlp to be merged first

### TLDR
`config.ClusterConfig` has 2 new fields:
* `ClusterConfig.EnvironmentPath`
* `ClusterConfig.EnsembleServiceK8sConfig`

### Current status
* Github workflows are passing
* Local dev set up changes still WIP

### CODE REVIEW

File: api/turing/cluster/controller.go [hunk 1]
```diff
@@ -165,29 +162,22 @@ func newController(clusterCfg clusterConfig) (Controller, error) {
...
-				return nil, errors.Wrapf(err,
-					"unable to get cluster secret for cluster: %s", clusterName)
-			}
+			creds := mlpcluster.NewK8sClusterCreds(k)
 
-			clusterCfg.Host = clusterSecret.Endpoint
-			clusterCfg.CACert = clusterSecret.CaCert
-			clusterCfg.ClientCert = clusterSecret.ClientCert
-			clusterCfg.ClientKey = clusterSecret.ClientKey
+			clusterCfg.Host = k.Cluster.Server
```
Comment (mbruner) [inline api/turing/cluster/controller.go:179]: The `Host` can be removed here. It is won't be used in `newController()` anymore

File: api/turing/config/config.go [hunk 2]
```diff
@@ -332,6 +346,24 @@ type ClusterConfig struct {
 	InClusterConfig bool
 	// VaultConfig is required if InClusterConfig is false.
 	VaultConfig *VaultConfig `validate:"required_without=InClusterConfig"`
```
Comment (mbruner) [inline api/turing/config/config.go:348]: Should be removed too

File: api/turing/api/appcontext.go [hunk 3]
```diff
@@ -190,20 +189,26 @@ func NewAppContext(
 // getEnvironmentClusterMap creates a map of the environment name to the kubernetes cluster. Additionally,
```
Comment (mbruner) [inline api/turing/api/appcontext.go:189]: This comment still refers to the old method

File: api/turing/api/appcontext.go [hunk 4]
```diff
@@ -128,7 +126,7 @@ func NewAppContext(
 
 		batchClusterController, ok := clusterControllers[cfg.BatchEnsemblingConfig.JobConfig.DefaultEnvironment]
 		if !ok {
-			return nil, errors.Wrapf(err, "Failed getting the batch ensembling job controller")
+			return nil, fmt.Errorf("Failed getting the batch ensembling job controller")
```
Comment (mbruner) [inline api/turing/api/appcontext.go:129]: Why switch from the wrapped error?

File: api/turing/api/appcontext.go [hunk 5]
```diff
@@ -62,18 +61,17 @@ func NewAppContext(
 		return nil, errors.Wrapf(err, "Failed initializing MLP Service")
 	}
 
-	// Create a map of env name to cluster name for each supported deployment environment
-	envClusterMap, err := getEnvironmentClusterMap(mlpSvc, []string{cfg.EnsemblerServiceBuilderConfig.ClusterName})
+	envClusterMap, err := buildKubeconfigStore(mlpSvc, cfg)
 	if err != nil {
-		return nil, errors.Wrapf(err, "Error obtaining environment info from MLP Service")
+		return nil, errors.Wrapf(err, "Error obtaining environment info from MLP Service and constructing kubeconfig store")
```
Comment (mbruner) [inline api/turing/api/appcontext.go:66]: Do we still obtain something from MLP, or do we rely on file only and mention MLP can be removed?

File: api/turing/api/appcontext.go [hunk 6]
```diff
@@ -149,6 +147,7 @@ func NewAppContext(
 		)
 		batchJobRunners = append(batchJobRunners, batchEnsemblingJobRunner)
 	}
+	fmt.Println("here")
```
Comment (mbruner) [inline api/turing/api/appcontext.go:150]: leftover debug code

File: api/turing/cluster/controller.go (hunk 1 above)
Comment (mbruner) [reply api/turing/cluster/controller.go:179]: Actually, `clusterConfig` structure will lose most of its fields

File: api/turing/api/appcontext.go (hunk 4 above)
Comment (shydefoo) [reply api/turing/api/appcontext.go:129]: err is actually nil if the code reaches here

File: api/turing/api/appcontext.go (hunk 5 above)
Comment (shydefoo) [reply api/turing/api/appcontext.go:66]: Yes, the list of environments is obtained from mlp first, before checking if there are k8sconfigs provided for those environments.

File: api/turing/api/appcontext.go (hunk 6 above)
Comment (shydefoo) [reply api/turing/api/appcontext.go:150]: my bad, removed.

File: api/turing/config/config.go (hunk 2 above)
Comment (shydefoo) [reply api/turing/config/config.go:348]: Yep, will remove this together with unit tests changes, as this field is still referenced there.

File: api/turing/api/appcontext.go (hunk 3 above)
Comment (shydefoo) [reply api/turing/api/appcontext.go]: Fixed

File: api/go.mod [hunk 7]
```diff
@@ -237,6 +233,7 @@ replace (
 	k8s.io/controller-manager => k8s.io/controller-manager v0.22.7
 	k8s.io/cri-api => k8s.io/cri-api v0.22.7
 	k8s.io/csi-translation-lib => k8s.io/csi-translation-lib v0.22.7
+	k8s.io/klog/v2 => k8s.io/klog/v2 v2.9.0
```
Comment (krithika369) [inline api/go.mod:236]: Where is this used?

**Review by krithika369 — COMMENTED** [review]
General note: The practice we generally adopt is to fork the repo rather than to create a branch on the original - this way, the main repo's branches can be cleaner, with the only branches being feature branches / for back porting.

File: api/go.mod (hunk 7 above)
Comment (shydefoo) [reply api/go.mod:236]: It's used as an indirect dependency in [mlp]( https://github.com/gojek/mlp/blob/05ede139570e1a9fc271fb03dca3fcfe7deb6074/go.mod#L113).

File: api/turing/api/appcontext.go [hunk 8]
```diff
@@ -187,23 +186,33 @@ func NewAppContext(
...
 }
 
-// getEnvironmentClusterMap creates a map of the environment name to the kubernetes cluster. Additionally,
+// buildKubeconfigStore creates a map of the environment name to the kubernetes cluster. Additionally,
 // clusters that are not a part of the deployment environments can be registered using the clusterNames
 // parameter (in such cases, the environment name will be saved to be the same as the cluster name).
-func getEnvironmentClusterMap(mlpSvc service.MLPService, clusterNames []string) (map[string]string, error) {
-	envClusterMap := map[string]string{}
+func buildKubeconfigStore(mlpSvc service.MLPService, cfg *config.Config) (map[string]*mlpcluster.K8sConfig, error) {
+
```
[the remark this answers is not in the document]
Comment (shydefoo) [reply api/turing/api/appcontext.go:193]: removed

**Comment by shydefoo** [comment]
Closing this PR first, opening from forked repo.

**Comment by shydefoo** [comment]
New PR opened here: https://github.com/caraml-dev/turing/pull/314

**Pull request closed without merging**